scripts/curate_data_axondeepseg_tem.py

```diff
@@ -0,0 +1,162 @@
+    os.system("find " + root_data + " -name '.DS_Store' -type f -delete")
+
+    # Remove output_data if exists to start clean
+    if os.path.isdir(output_data):
+        shutil.rmtree(output_data)
+    os.makedirs(output_data, exist_ok=True)
+
+    contents_ds = [subdir for subdir in os.listdir(root_data) if os.path.isdir(os.path.join(root_data, subdir))]
+    contents_ds.sort()
+
+    # Loop across contents of each subdirectory
+    for subdir in contents_ds:
+        sub_id = "sub-nyuMouse" + subdir.split('_')[3]
+        sample_id = subdir.split('_')[4]
+        path_subdir = os.path.join(root_data, subdir)
+        contents_subdir = os.listdir(path_subdir)
+        sub_bids_full = sub_id + "_sample-" + sample_id
+        for file in contents_subdir:
+            path_file_in = os.path.join(path_subdir, file)
+            if file in images:
+                path_sub_id_dir_out = os.path.join(output_data, sub_id, 'microscopy')
+                path_file_out = os.path.join(path_sub_id_dir_out, sub_bids_full + images[file])
+                create_path_sub_id_dir_out(path_sub_id_dir_out)
+                create_json_sidecar(path_sub_id_dir_out, sub_id + '_TEM.png')
+            if file in der:
+                path_sub_id_dir_out = os.path.join(output_data, 'derivatives', 'labels', sub_id, 'microscopy')
+                path_file_out = os.path.join(path_sub_id_dir_out, sub_bids_full + der[file])
+                create_path_sub_id_dir_out(path_sub_id_dir_out)
+            shutil.copy(path_file_in, path_file_out)
+
+    sub_list = os.listdir(output_data)
+    sub_list.remove('derivatives')
+    sub_list.sort()
+
+    # Write additional files
+    import csv
+    participants = []
+    samples = []
+    for subject in sub_list:
+        row_sub = []
+        row_sub.append(subject)
+        participants.append(row_sub)
+        list_samples = []
+        for sample in os.listdir(os.path.join(output_data, subject, 'microscopy')):
+            if sample.endswith('.png'):
+                list_samples.append(sample)
+        list_samples.sort()
+        for file_sample in list_samples:
+            row_sub_samples = []
+            row_sub_samples.append(subject)
+            row_sub_samples.append(file_sample.split('_')[1])
+            row_sub_samples.append('tissue')
+            samples.append(row_sub_samples)
+
+    # Create participants.tsv
+    with open(output_data + '/participants.tsv', 'w') as tsv_file:
+        tsv_writer = csv.writer(tsv_file, delimiter='\t', lineterminator='\n')
+        tsv_writer.writerow(["participant_id"])
+        for item in participants:
+            tsv_writer.writerow(item)
+
+    # Create samples.tsv
+    with open(output_data + '/samples.tsv', 'w') as tsv_file:
+        tsv_writer = csv.writer(tsv_file, delimiter='\t', lineterminator='\n')
+        tsv_writer.writerow(["participant_id", "sample_id", "sample_type"])
+        for item in samples:
+            tsv_writer.writerow(item)
```

Comment:
You should be able to combine this loop and the loop that creates `samples`. If you do that, you can avoid buffering the entire dataset in memory: instead of creating `samples`, just create `row_sub_samples` and immediately write it out.

Move the `with open():` above the above loop, and indent that loop inside of it, then replace `for item in samples` with your `for subject in sub_list:`. Everything else can be probably kept the same :)

Reply:
Oh but it is a little bit more complicated because you're turning one list `sub_list` into two `participants` and `samples`. It's still possible to rewrite this in a streaming way though:

```
     with open(output_data + '/samples.tsv', 'w') as samples, \
          open(output_data + '/participants.tsv', 'w') as participants:
        for subject in sub_list:
            ....    
            participants.writerow(row_sub)
            ....
            for sample in subject_samples:
                ...
                samples.writerow(row_sub_samples)
```